frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -867,30 +861,24 @@ TypeInference::checkExternConstructor(const IR::Node* errorPosition,
             paramType->is<IR::Type_Package>())
             typeError("%1%: parameter cannot have type %2%", pi, paramType);
 
-        auto tvs = unify(errorPosition, paramType, argType);
-        if (tvs == nullptr) {
-            // error already signalled
-            return nullptr;
-        }
-        if (tvs->isIdentity()) {
-            result->push_back(arg);
-            continue;
-        }
+        auto argInfo = new IR::ArgumentInfo(arg->srcInfo, arg->expression, true, argType, arg);
+        args->push_back(argInfo);
+    }
 
-        ConstantTypeSubstitution cts(tvs, refMap, typeMap, this);
-        auto newArg = cts.convert(arg->expression);
-        if (::errorCount() > 0)
-            return arguments;
+    auto rettype = new IR::Type_Var(IR::ID(refMap->newName("R")));
+    auto callType = new IR::Type_MethodCall(new IR::Vector<IR::Type>(),
+                                            rettype, args);
+    TypeConstraints constraints(typeMap->getSubstitutions());
+    constraints.addEqualityConstraint(mt, callType);
+    auto tvs = constraints.solve(errorPosition);
+    BUG_CHECK(tvs != nullptr || ::errorCount(), "Null substitution");
+    if (tvs == nullptr)
+        return nullptr;
+    addSubstitutions(tvs);
 
-        result->push_back(new IR::Argument(arg->srcInfo, arg->name, newArg));
-        setType(newArg, paramType);
-        if (newArg != arg->expression)
-            changes = true;
-    }
-    if (changes)
-        return result;
-    else
-        return arguments;
```

Comment:
Looks like the code is no longer checking for "no changes" and returning the original call?

Reply:
Yes, all changes will be made in the cts.convert call. There's nothing else left for us to track.